velox/expression/Expr.cpp

```diff
@@ -375,46 +376,117 @@ std::optional<std::string> generateFilePath(
 }
 } // namespace
 
-void ExprExceptionContext::persistDataAndSql(const char* basePath) {
-  // Exception already persisted or failed to persist. We don't persist again in
-  // this situation.
-  if (!dataPath_.empty()) {
-    return;
-  }
-
+// static
+bool ExprSaver::saveData(
+    const BaseVector* FOLLY_NONNULL vector,
+    const char* FOLLY_NONNULL basePath,
+    std::string& storedPath,
+    std::string& failMsg) {
   auto dataPath = generateFilePath(basePath, "vector");
-  auto sqlPath = generateFilePath(basePath, "sql");
   if (!dataPath.has_value()) {
-    dataPath_ = "Failed to create file for saving input vector.";
-    return;
-  }
-  if (!sqlPath.has_value()) {
-    sqlPath_ = "Failed to create file for saving expression SQL.";
-    return;
+    failMsg = "Failed to create file for saving input vector.";
+    return false;
   }
 
   // Persist vector to disk
   try {
     std::ofstream outputFile(dataPath.value(), std::ofstream::binary);
-    saveVector(*vector_, outputFile);
+    saveVector(*vector, outputFile);
     outputFile.close();
-    dataPath_ = dataPath.value();
+    storedPath = dataPath.value();
+  } catch (std::exception& e) {
+    failMsg = fmt::format(
+        "Failed to save input vector to {}. Reasons: '{}'",
+        dataPath.value(),
+        e.what());
+    return false;
   } catch (...) {
-    dataPath_ =
+    failMsg =
         fmt::format("Failed to save input vector to {}.", dataPath.value());
+    return false;
+  }
+  return true;
+}
+
+// static
+bool ExprSaver::saveSql(
+    const std::string& sql,
+    const char* FOLLY_NONNULL basePath,
+    std::string& storedPath,
+    std::string& failMsg) {
+  auto sqlPath = generateFilePath(basePath, "sql");
+  if (!sqlPath.has_value()) {
+    failMsg = "Failed to create file for saving expression SQL.";
+    return false;
   }
 
   // Persist sql to disk
   try {
-    auto sql = expr_->toSql();
     std::ofstream outputFile(sqlPath.value(), std::ofstream::binary);
     outputFile.write(sql.data(), sql.size());
     outputFile.close();
-    sqlPath_ = sqlPath.value();
+    storedPath = sqlPath.value();
+  } catch (std::exception& e) {
+    failMsg = fmt::format(
+        "Failed to save expression SQL to {}. Reasons: '{}'",
+        sqlPath.value(),
+        e.what());
+    return false;
   } catch (...) {
-    sqlPath_ =
+    failMsg =
         fmt::format("Failed to save expression SQL to {}.", sqlPath.value());
+    return false;
+  }
+  return true;
+}
+
+void ExprExceptionContext::persistDataAndSql(const char* basePath) {
```

Comment:
Did you end up not using ExprExceptionContext in the Fuzzer? Should we move it back into the .cpp file?

Reply:
Right I separated the persistence logic out. Any reason we want to move it back?